test/git-mirage/test.ml

```diff
@@ -133,9 +133,28 @@ module TCP = Test_sync.Make(struct
     module M = Git_mirage.Sync(MirageConduit)(Git.Mem.Store(Digestif.SHA1))
     module Store = M.Store
     type error = M.error
-    let clone t ~reference uri = M.clone t ~reference uri
-    let fetch_all t uri = M.fetch_all t uri
-    let update t ~reference uri = M.update t ~reference uri
+    let clone t ~reference uri = M.clone t ~reference:(reference, reference) uri
+
+    exception Jump of Store.Ref.error
+
+    let fetch_all t uri =
+      let open Lwt.Infix in
+
+      M.fetch_all t ~references:Store.Reference.Map.empty uri >>= function
+      | Error _ as err -> Lwt.return err
+      | Ok (_, _, downloaded) ->
+        Lwt.try_bind
+          (fun () -> Lwt_list.iter_s
+              (fun (remote_ref, new_hash) ->
+                 Store.Ref.write t remote_ref (Store.Reference.Hash new_hash) >>= function
+                 | Ok () -> Lwt.return ()
+                 | Error err -> Lwt.fail (Jump err))
+              (Store.Reference.Map.bindings downloaded))
+          (fun () -> Lwt.return (Ok ()))
+          (function Jump err -> Lwt. return (Error (`Ref err))
+                  | err -> Lwt.fail err)
```

Comment:
all of this will probably be copy/pasted by users of the TCP sync. So this should probably go into the `Sync` module ...

Reply:
About this part of the PR is a side effect of the new API in 2 terms (the bottom and the top).

At the bottom, it's only about API.

For me, `fetch_all` should not automatically create the references itself. It creates references only according to client wishes (by the arguments `~references`).

However, `fetch_all` downloads all references and if the references argument is not exhaustive (and, I believe it's a common case), the client could lost some useful hashes endpoints (because they are not saved in references).

Before this patch, I provided this signature:

```ocaml
val fetch_all:
   Store.t -> ?locks:Store.Lock.t ->
   ?capabilities:Capability.t list ->
   create:(Store.Reference.t -> Store.Reference.t option Lwt.t) ->
   references:Store.Reference.t list Store.Reference.Map.t ->
   Uri.t -> (Store.Hash.t Store.Reference.Map.t
             * Store.Reference.t list Store.Reference.Map.t
             * Store.Hash.t Store.Reference.Map.t, error) result Lwt.t
```

As you can see, we expect a `create` function which is a process to make a new local reference from the already downloaded remote reference. Then, we save the hash endpoint to this new local reference (which does not appear in references).

I'm not very happy about this previous signature for two reasons:
- firstly, we complexify a lot the API - we need to explain to the client: "ok, all references in the references will surely be created but the create function will create local references too"
- secondly, I don't like to let the user to define a black box function

And because I don't like to let `fetch_all` to create arbitrary local references, I prefer to have this current API.

At the top, it's only about what the test expect.

When I did this patch, I thinked to keep the previous same semantic about `fetch_all` and I did not think if it's necessary to have systematically all local references binded with remote references - I did not look where this function was used and if it's necessary to have this semantic.

If I remember correctly, the test on the `fetch_all` works only on one reference, master. It could be (may be?) more useful to provides `~references` as an argument of the `fetch_all` function available on the [SYNC](https://github.com/mirage/ocaml-git/blob/master/test/common/test_sync.ml#L21) signature and explicitly want to save the remote master on the local master.